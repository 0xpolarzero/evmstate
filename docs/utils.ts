import { isHex } from "tevm";
import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const maxBytesLength = 64;
export const stringify = (value: any, space: number = 2) => {
  return JSON.stringify(
    value,
    (_, v) => {
      if (typeof v === "bigint") return `__bigint__${v.toString()}`;
      if (isHex(v) && (v.length - 2) / 2 > maxBytesLength)
        return `__long_hex__${v.slice(0, maxBytesLength + 2)}...${v.slice(-maxBytesLength)}`;
      return v;
    },
    space,
  );
};
