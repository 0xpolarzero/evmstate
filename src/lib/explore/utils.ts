import { Hex, toHex } from "tevm";
import { ByteArray, ToHexParameters } from "viem";

export const max = <T extends bigint | number>(...nums: Array<T>): T =>
  // @ts-expect-error bigint/number
  nums.reduce((max, num) => (num > max ? num : max), -Infinity);

/**
 * Converts a value to a hex string and ensures it has an even number of characters (full bytes)
 *
 * @param value The value to convert to hex
 * @param options Optional configuration options passed to tevm's toHex
 * @returns A hex string with full bytes
 */
export const toHexFullBytes = (value: string | number | bigint | boolean | ByteArray, opts?: ToHexParameters): Hex => {
  // Use tevm's toHex function
  const hex = toHex(value, opts);

  // If the hex string has an odd number of characters after the 0x prefix,
  // pad it with a leading zero to make it even
  if ((hex.length - 2) % 2 !== 0) return ("0x0" + hex.slice(2)) as Hex;
  return hex;
};
