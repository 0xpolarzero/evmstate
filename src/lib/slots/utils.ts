import { Address, decodeAbiParameters, encodeAbiParameters, Hex, isHex, toHex } from "tevm";
import { abi } from "@shazow/whatsabi";
import { AbiType, AbiTypeToPrimitiveType } from "abitype";
import { padHex } from "viem";

import { MappingKey } from "@/lib/slots/types";

/* -------------------------------------------------------------------------- */
/*                                  MAPPINGS                                  */
/* -------------------------------------------------------------------------- */

/** Extract values from a transaction trace that might be used as mapping keys */
// TODO: refactor to navigate the stack trace and identify both potential keys and array indexes
/**
 * Typical access patterns:
 *
 * Mapping
 *
 * - KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key, uint256(mappingSlot)))
 * - -> key is the input to the hash so we might even be able to retrieve the slot and compare directly
 * - KECCAK256 → KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key2, keccak256(abi.encode(key1, uint256(mappingSlot)))))
 * - -> same here we get each key from each keccak input
 *
 * Array
 *
 * - KECCAK256 → ADD → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(uint256(arraySlot)) + index
 *
 * As reference:
 *
 * - Struct in array: KECCAK256 → ADD → ADD → SLOAD/SSTORE Slot calculation: (keccak256(uint256(arraySlot)) + index) +
 *   memberOffset
 * - Mapping to Struct with Mapping: KECCAK256 → ADD → KECCAK256 → SLOAD/SSTORE Slot calculation:
 *   keccak256(abi.encode(key2, (keccak256(abi.encode(key1, uint256(mappingSlot))) + memberOffset)))
 */
export const extractPotentialKeys = (
  trace: {
    uniqueStackValues?: Array<string>;
    relevantOps?: Array<{
      op: string;
      stack: Array<string>;
    }>;
  },
  addresses: Array<Address>,
  abiFunctions: Array<abi.ABIFunction>,
  txData?: Hex,
): MappingKey[] => {
  const keys: MappingKey[] = [];

  // Add touched addresses
  addresses.forEach((address) => {
    keys.push({
      hex: padHex(address, { size: 32 }),
      decoded: address,
      type: "address",
    });
  });

  // Extract parameters from transaction data
  if (txData && txData.length > 10) {
    const selector = txData.slice(0, 10);
    const inputs = abiFunctions.find((fn) => fn.selector === selector)?.inputs;

    if (inputs) {
      // Decode function inputs
      const params = decodeAbiParameters(inputs, `0x${txData.slice(10)}`);

      params.forEach((param, index) => {
        // If it's an array, add each element as a key
        if (Array.isArray(param)) {
          param.forEach((p) => {
            const type = inputs[index].type.replace("[]", "") as AbiType;

            if (type && type !== "string" && type !== "bytes") {
              keys.push({
                hex: padHex(encodeAbiParameters([{ type }], [p]), { size: 32 }),
                decoded: p as AbiTypeToPrimitiveType<typeof type>,
                type,
              });
            }
          });
        } else {
          // Otherwise just add the key straight up
          const type = inputs[index].type as AbiType;

          if (type && type !== "string" && type !== "bytes") {
            keys.push({
              hex: padHex(encodeAbiParameters([{ type }], [param]), { size: 32 }),
              decoded: param as AbiTypeToPrimitiveType<typeof type>,
              type,
            });
          }
        }
      });
    }
  }

  // Process stack values from the trace
  if (trace.uniqueStackValues?.length) {
    // Process unique stack values directly
    for (const stackValue of trace.uniqueStackValues) {
      keys.push({
        hex: isHex(stackValue) ? padHex(stackValue, { size: 32 }) : toHex(stackValue, { size: 32 }),
        type: undefined,
      });
    }
  }

  // Deduplicate keys
  const uniqueMap = new Map();
  // Add the new key only if it's not already in the map (and don't replace a key with a defined type)
  keys.forEach((k) => {
    if (!uniqueMap.has(k.hex) || k.type) uniqueMap.set(k.hex, k);
  });

  return Array.from(uniqueMap.values()).sort((a, b) => {
    // prefer address as it's more likely to be a key
    if (a.type === "address") return -1;
    if (b.type === "address") return 1;
    // prefer defined types
    if (a.type === undefined) return 1;
    if (b.type === undefined) return -1;

    return 0;
  });
};

/* -------------------------------------------------------------------------- */
/*                                    UTILS                                   */
/* -------------------------------------------------------------------------- */

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
