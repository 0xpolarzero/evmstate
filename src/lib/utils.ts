import { Address } from "tevm";

export const uniqueAddresses = (addresses: Array<Address | undefined>): Array<Address> => {
  let existingAddresses = new Set<string>();

  return addresses.filter((address) => {
    if (!address || existingAddresses.has(address.toLowerCase())) return false;
    existingAddresses.add(address.toLowerCase());
    return true;
  }) as Address[];
};
