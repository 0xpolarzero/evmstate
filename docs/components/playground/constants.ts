import { EthjsAddress } from "tevm/utils";
import type { ExtractAbiFunctions } from "abitype";

import * as LAYOUTS from "@test/codegen/layouts/index.js";
import * as CONTRACTS from "@test/contracts/index.js";

export const contract = CONTRACTS["Playground"].withAddress(
  EthjsAddress.fromString("0x987C2AF139EAEaBdF8D6d3d1723C1883bEa1f2AF").toString(),
);
export const layout = LAYOUTS["Playground"];
export const callerAddress = EthjsAddress.fromString("0xCa11e40000000000000000000000000000000000");
export const localStorageKey = "EVMSTATE_PLAYGROUND_STATE";

export const functionDescriptions: Record<ExtractAbiFunctions<typeof contract.abi, "nonpayable">["name"], string> = {
  addValue: "Adds to a dynamic array",
  addUser: "Adds a user struct to a mapping and array",
  toggleUserActive: "Toggles a boolean in a struct within a mapping",
  setBalance: "Updates a simple mapping value",
  setAllowance: "Updates a nested mapping value",
  addTransaction: "Adds to a dynamic array within a mapping",
  updateBasicValues: "Updates primitive types (uint256, bool)",
  updatePackedValues: "Updates packed storage variables (uint8, uint16, uint32, bool)",
  setStringAndBytes: "Updates string and bytes storage",
  setFixedValue: "Updates a fixed-size array element",
};
