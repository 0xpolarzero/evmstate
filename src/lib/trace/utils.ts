import {
  Abi,
  BlockTag,
  ContractFunctionName,
  createMemoryClient,
  encodeFunctionData,
  Hex,
  http,
  MemoryClient,
} from "tevm";
import { Common } from "tevm/common";

import { debug } from "@/debug";
import { ExploreStorageConfig } from "@/lib/explore/config";
import { TraceStorageAccessOptions, TraceStorageAccessTxParams, TraceStorageAccessTxWithData } from "@/lib/trace/types";

/** Creates a Tevm client from the provided options */
export const createClient = (options: {
  rpcUrl?: string;
  common?: Common;
  blockTag?: BlockTag | bigint;
}): MemoryClient => {
  const { rpcUrl, common, blockTag } = options;
  if (!rpcUrl) throw new Error("You need to provide a rpcUrl if you don't provide a client directly");

  return createMemoryClient({
    common,
    fork: {
      transport: http(rpcUrl),
      blockTag: blockTag ?? "latest",
    },
    miningConfig: { type: "manual" },
  });
};

export const getUnifiedParams = async <
  TAbi extends Abi | readonly unknown[] = Abi,
  TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
>(
  args: TraceStorageAccessOptions & TraceStorageAccessTxParams<TAbi, TFunctionName> & { config?: ExploreStorageConfig },
): Promise<TraceStorageAccessTxWithData & { client: MemoryClient }> => {
  const { client: _client, rpcUrl, common } = args;

  // Create the tevm client
  const client = _client ?? createClient({ rpcUrl, common });

  // Return early if the tx was already provided in calldata format
  if (args.from && args.data) return { client, from: args.from, to: args.to, data: args.data, value: args.value };

  // Encode calldata if the contract call was provided (abi, functionName, args)
  if (args.from && args.to && args.abi && args.functionName && args.args) {
    try {
      // @ts-expect-error complex union type not exactly similar
      const data = encodeFunctionData(args);
      return { client, from: args.from, to: args.to, data, value: args.value };
    } catch (err) {
      debug(`Failed to encode function data: ${err}`);
      throw err;
    }
  }

  // If we provide a value but no txHash, it's a simple transfer
  if (args.value && !args.txHash) return { client, from: args.from, to: args.to, value: args.value };

  // In this case, we need to replay the transaction
  if (!args.txHash)
    throw new Error("You need to provide a txHash if you don't provide the transaction data or contract call");

  // If we're replaying a transaction, extract the from, to, and data from the transaction
  try {
    const tx = await client.getTransaction({ hash: args.txHash });

    // Duplicate the client but fork before the transaction's block
    const clientBeforeTx = createMemoryClient({
      ...client,
      fork: {
        transport: client.transport,
        // Ideally we'd want to fork just right before that tx as some same-block txs can impact the state
        blockTag: tx.blockNumber > 0 ? tx.blockNumber - BigInt(1) : BigInt(0),
      },
    });

    return {
      client: clientBeforeTx,
      from: tx.from,
      to: tx.to ?? undefined,
      // TODO: remove when correctly formatted (tx in block mined here has data instead of input)
      // @ts-expect-error Property 'data' does not exist on type Transaction
      data: tx.input ? (tx.input as Hex) : (tx.data as Hex),
      value: tx.value,
    };
  } catch (err) {
    debug(`Failed to get transaction for replaying ${args.txHash}: ${err}`);
    throw err;
  }
};

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
