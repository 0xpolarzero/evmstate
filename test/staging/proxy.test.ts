import { createMemoryClient, encodeFunctionData, Hex, http } from "tevm";
import { createAddress } from "tevm/address";
import { EthjsAddress } from "tevm/utils";
import { assert, beforeEach, describe, expect, it } from "vitest";

import { FORK, LAYOUTS } from "@test/constants";
import { solidityDebugger } from "@test/debugger";
import { expectedStorage, getSlotHex, toEvenHex } from "@test/utils";
import { traceStorageAccess } from "@/index";

const { TransparentProxy, CounterImpl, CounterImplV2 } = FORK.mainnet.contracts;

const mainnet = FORK.mainnet;
const client = createMemoryClient({
  common: mainnet.common,
  fork: {
    transport: http(mainnet.rpcUrl),
    blockTag: "latest",
  },
});

describe("Proxies", () => {
  let admin: EthjsAddress;

  beforeEach(async () => {
    const res = await client.tevmContract(TransparentProxy.read.getAdmin());
    assert(res.data, "Could not get proxy admin");
    admin = createAddress(res.data);
  });

  it("should trace storage changes through a proxy", async () => {
    const trace = await traceStorageAccess({
      client,
      from: admin.toString(),
      to: TransparentProxy.address,
      abi: CounterImpl.abi,
      functionName: "setCount",
      args: [100n],
      explorers: mainnet.explorers,
    });

    expect(trace[TransparentProxy.address.toLowerCase() as Hex].storage).toEqual(
      expectedStorage(LAYOUTS.CounterImpl, {
        _count: {
          name: "_count",
          type: "uint256",
          kind: "primitive",
          trace: [
            {
              current: { hex: toEvenHex(0, { size: 32 }), decoded: 0n },
              next: { hex: toEvenHex(100, { size: 32 }), decoded: 100n },
              modified: true,
              slots: [getSlotHex(0)],
              path: [],
              fullExpression: "_count",
            },
          ],
        },
        // @ts-expect-error __implementation field does not exist
        __implementation: {
          name: "__implementation",
          type: "address",
          kind: "primitive",
          trace: [
            {
              current: { hex: CounterImpl.address.toLowerCase(), decoded: CounterImpl.address },
              modified: false,
              slots: ["0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"],
              path: [],
              fullExpression: "__implementation",
            },
          ],
        },
      }),
    );
  });

  it("should track proxy-specific storage slots", async () => {
    const trace = await traceStorageAccess({
      client,
      from: admin.toString(),
      to: TransparentProxy.address,
      abi: TransparentProxy.abi,
      functionName: "changeImplementation",
      args: [CounterImplV2.address],
      explorers: mainnet.explorers,
    });

    expect(trace[TransparentProxy.address.toLowerCase() as Hex].storage).toEqual(
      expectedStorage(LAYOUTS.CounterImpl, {
        // @ts-expect-error __implementation field does not exist
        __implementation: {
          name: "__implementation",
          type: "address",
          kind: "primitive",
          trace: [
            {
              current: { hex: CounterImpl.address.toLowerCase(), decoded: CounterImpl.address },
              next: { hex: CounterImplV2.address.toLowerCase(), decoded: CounterImplV2.address },
              modified: true,
              slots: ["0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"],
              path: [],
              fullExpression: "__implementation",
            },
          ],
        },
        __admin: {
          name: "__admin",
          type: "address",
          kind: "primitive",
          trace: [
            {
              current: { hex: admin.toString().toLowerCase(), decoded: admin.toString() },
              modified: false,
              slots: ["0xb53127684a568b3173ae13b9f8a6016e019b2c8e8cbb2a6e0a23387fdaa12345"],
              path: [],
              fullExpression: "__admin",
            },
          ],
        },
      }),
    );
  });

  it("should retrieve new storage layout when implementation is upgraded", async () => {
    // Update count
    await client.tevmCall({
      caller: admin.toString(),
      to: TransparentProxy.address,
      data: encodeFunctionData({
        abi: CounterImpl.abi,
        functionName: "setCount",
        args: [100n],
      }),
      addToBlockchain: true,
    });

    // Update implementation
    await client.tevmCall({
      from: admin.toString(),
      to: TransparentProxy.address,
      data: encodeFunctionData({
        abi: TransparentProxy.abi,
        functionName: "changeImplementation",
        args: [CounterImplV2.address],
      }),
      addToBlockchain: true,
    });

    const trace = await traceStorageAccess({
      client,
      from: admin.toString(),
      to: TransparentProxy.address,
      abi: CounterImplV2.abi,
      functionName: "setCount",
      args: [200n],
      explorers: mainnet.explorers,
    });

    expect(trace[TransparentProxy.address.toLowerCase() as Hex].storage).toEqual(
      expectedStorage(LAYOUTS.CounterImplV2, {
        _count: {
          name: "_count",
          type: "uint256",
          kind: "primitive",
          trace: [
            {
              current: { hex: toEvenHex(100, { size: 32 }), decoded: 100n },
              next: { hex: toEvenHex(200, { size: 32 }), decoded: 200n },
              modified: true,
              slots: [getSlotHex(0)],
              path: [],
              fullExpression: "_count",
            },
          ],
        },
        _paused: {
          name: "_paused",
          type: "bool",
          kind: "primitive",
          trace: [
            {
              current: { hex: toEvenHex(0), decoded: false },
              modified: false,
              slots: [getSlotHex(1)],
              path: [],
              fullExpression: "_paused",
            },
          ],
        },
        // @ts-expect-error __implementation field does not exist
        __implementation: {
          name: "__implementation",
          type: "address",
          kind: "primitive",
          trace: [
            {
              current: { hex: CounterImplV2.address.toLowerCase(), decoded: CounterImplV2.address },
              modified: false,
              slots: ["0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"],
              path: [],
              fullExpression: "__implementation",
            },
          ],
        },
      }),
    );
  });
});
