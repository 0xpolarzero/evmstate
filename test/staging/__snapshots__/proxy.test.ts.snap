// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Proxies > should trace storage changes through a proxy 1`] = `
Map {
  "0xc22b2d4f82b599f64f93c0185a044245a64b8c42" => {
    "balance": {
      "current": 1000000000000000000n,
      "modified": true,
      "next": 999999999999658848n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 202,
      "modified": true,
      "next": 203,
    },
    "storage": {},
  },
  "0x6f64746625f57a146a8f2168852d73dfec6771e8" => {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x608060405260043610610037575f3560e01c806317a68dd8146100d75780636e9960c3146100ff578063aaf10f42146101295761003e565b3661003e57005b5f610047610153565b90505f73ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff16036100b7576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016100ae90610343565b60405180910390fd5b365f80375f80365f845af43d805f803e815f81146100d357815ff35b815ffd5b3480156100e2575f80fd5b506100fd60048036038101906100f891906103bf565b610182565b005b34801561010a575f80fd5b50610113610203565b60405161012091906103f9565b60405180910390f35b348015610134575f80fd5b5061013d610211565b60405161014a91906103f9565b60405180910390f35b5f807f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc5f1b9050805491505090565b61018a61021f565b73ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff16146101f7576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016101ee90610482565b60405180910390fd5b6102008161024e565b50565b5f61020c61021f565b905090565b5f61021a610153565b905090565b5f807fb53127684a568b3173ae13b9f8a6016e019b2c8e8cbb2a6e0a23387fdaa123455f1b9050805491505090565b5f73ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff16036102bc576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102b3906104ea565b60405180910390fd5b5f7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc5f1b90508181555050565b5f82825260208201905092915050565b7f496d706c656d656e746174696f6e206e6f7420736574000000000000000000005f82015250565b5f61032d6016836102e9565b9150610338826102f9565b602082019050919050565b5f6020820190508181035f83015261035a81610321565b9050919050565b5f80fd5b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f61038e82610365565b9050919050565b61039e81610384565b81146103a8575f80fd5b50565b5f813590506103b981610395565b92915050565b5f602082840312156103d4576103d3610361565b5b5f6103e1848285016103ab565b91505092915050565b6103f381610384565b82525050565b5f60208201905061040c5f8301846103ea565b92915050565b7f4f6e6c792061646d696e2063616e206368616e676520696d706c656d656e74615f8201527f74696f6e00000000000000000000000000000000000000000000000000000000602082015250565b5f61046c6024836102e9565b915061047782610412565b604082019050919050565b5f6020820190508181035f83015261049981610460565b9050919050565b7f496e76616c696420696d706c656d656e746174696f6e000000000000000000005f82015250565b5f6104d46016836102e9565b91506104df826104a0565b602082019050919050565b5f6020820190508181035f830152610501816104c8565b905091905056fea2646970667358221220ee487cb9c74cf1226102f2d821f8d835876ae0008226f5dcaaeac35493322bbf64736f6c63430008170033",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": false,
    },
    "storage": {
      "__implementation": {
        "kind": "primitive",
        "name": "__implementation",
        "trace": [
          {
            "current": {
              "decoded": "0x47676f15E78BfFE13225eFF46C6A2BDd3199ed95",
              "hex": "0x47676f15e78bffe13225eff46c6a2bdd3199ed95",
            },
            "fullExpression": "__implementation",
            "modified": false,
            "path": [],
            "slots": [
              "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
            ],
          },
        ],
        "type": "address",
      },
      "_count": {
        "kind": "primitive",
        "name": "_count",
        "trace": [
          {
            "current": {
              "decoded": 0n,
              "hex": "0x0000000000000000000000000000000000000000000000000000000000000000",
            },
            "fullExpression": "_count",
            "modified": true,
            "next": {
              "decoded": 100n,
              "hex": "0x0000000000000000000000000000000000000000000000000000000000000064",
            },
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
        "type": "uint256",
      },
    },
  },
  "0x0000000000000000000000000000000000000000" => {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
}
`;

exports[`Proxies > should track proxy-specific storage slots 1`] = `
Map {
  "0xc22b2d4f82b599f64f93c0185a044245a64b8c42" => {
    "balance": {
      "current": 1000000000000000000n,
      "modified": true,
      "next": 999999999999796083n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 202,
      "modified": true,
      "next": 203,
    },
    "storage": {},
  },
  "0x6f64746625f57a146a8f2168852d73dfec6771e8" => {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x608060405260043610610037575f3560e01c806317a68dd8146100d75780636e9960c3146100ff578063aaf10f42146101295761003e565b3661003e57005b5f610047610153565b90505f73ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff16036100b7576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016100ae90610343565b60405180910390fd5b365f80375f80365f845af43d805f803e815f81146100d357815ff35b815ffd5b3480156100e2575f80fd5b506100fd60048036038101906100f891906103bf565b610182565b005b34801561010a575f80fd5b50610113610203565b60405161012091906103f9565b60405180910390f35b348015610134575f80fd5b5061013d610211565b60405161014a91906103f9565b60405180910390f35b5f807f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc5f1b9050805491505090565b61018a61021f565b73ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff16146101f7576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016101ee90610482565b60405180910390fd5b6102008161024e565b50565b5f61020c61021f565b905090565b5f61021a610153565b905090565b5f807fb53127684a568b3173ae13b9f8a6016e019b2c8e8cbb2a6e0a23387fdaa123455f1b9050805491505090565b5f73ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff16036102bc576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016102b3906104ea565b60405180910390fd5b5f7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc5f1b90508181555050565b5f82825260208201905092915050565b7f496d706c656d656e746174696f6e206e6f7420736574000000000000000000005f82015250565b5f61032d6016836102e9565b9150610338826102f9565b602082019050919050565b5f6020820190508181035f83015261035a81610321565b9050919050565b5f80fd5b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f61038e82610365565b9050919050565b61039e81610384565b81146103a8575f80fd5b50565b5f813590506103b981610395565b92915050565b5f602082840312156103d4576103d3610361565b5b5f6103e1848285016103ab565b91505092915050565b6103f381610384565b82525050565b5f60208201905061040c5f8301846103ea565b92915050565b7f4f6e6c792061646d696e2063616e206368616e676520696d706c656d656e74615f8201527f74696f6e00000000000000000000000000000000000000000000000000000000602082015250565b5f61046c6024836102e9565b915061047782610412565b604082019050919050565b5f6020820190508181035f83015261049981610460565b9050919050565b7f496e76616c696420696d706c656d656e746174696f6e000000000000000000005f82015250565b5f6104d46016836102e9565b91506104df826104a0565b602082019050919050565b5f6020820190508181035f830152610501816104c8565b905091905056fea2646970667358221220ee487cb9c74cf1226102f2d821f8d835876ae0008226f5dcaaeac35493322bbf64736f6c63430008170033",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": false,
    },
    "storage": {
      "__admin": {
        "kind": "primitive",
        "name": "__admin",
        "trace": [
          {
            "current": {
              "decoded": "0xc22B2d4F82b599f64F93c0185A044245A64b8C42",
              "hex": "0xc22b2d4f82b599f64f93c0185a044245a64b8c42",
            },
            "fullExpression": "__admin",
            "modified": false,
            "path": [],
            "slots": [
              "0xb53127684a568b3173ae13b9f8a6016e019b2c8e8cbb2a6e0a23387fdaa12345",
            ],
          },
        ],
        "type": "address",
      },
      "__implementation": {
        "kind": "primitive",
        "name": "__implementation",
        "trace": [
          {
            "current": {
              "decoded": "0x47676f15E78BfFE13225eFF46C6A2BDd3199ed95",
              "hex": "0x47676f15e78bffe13225eff46c6a2bdd3199ed95",
            },
            "fullExpression": "__implementation",
            "modified": true,
            "next": {
              "decoded": "0xCC0eAE411714557Ac10A0b3F310d17AA1C72A360",
              "hex": "0xcc0eae411714557ac10a0b3f310d17aa1c72a360",
            },
            "path": [],
            "slots": [
              "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc",
            ],
          },
        ],
        "type": "address",
      },
    },
  },
  "0x0000000000000000000000000000000000000000" => {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
}
`;
